### PR TITLE
`addr_of!` grants mutable access, maybe?

### PR DESCRIPTION
**Pull request by tmiasko**
The exact set of permissions granted when forming a raw reference is
currently undecided https://github.com/rust-lang/rust/issues/56604.

To avoid presupposing any particular outcome, adjust the const
qualification to be compatible with decision where raw reference
constructed from `addr_of!` grants mutable access.

Additionally, to avoid keeping `MaybeMutBorrowedLocals` in sync with
const qualification, remove it. It's no longer used.

@rust-lang/wg-const-eval

### CODE REVIEW

**Comment by tmiasko** [comment]
@oli-obk @ecstatic-morse to address concerns from #90214 about duplication with `MaybeMutBorrowedLocals`, I decided to remove it, since we no longer used it anywhere else, but let me know if you prefer to retain it.

**Comment by oli-obk** [comment]
Ah, if that was the only user, then this is perfect!

File: compiler/rustc_const_eval/src/transform/check_consts/resolver.rs [hunk 1]
```diff
@@ -265,6 +273,17 @@ pub(super) struct State {
     pub borrow: BitSet<Local>,
 }
 
+impl Clone for State {
```
Comment (oli-obk) [inline compiler/rustc_const_eval/src/transform/check_consts/resolver.rs:276]: Please put the comment from the commit on this impl, so future readers don't have to look at blame

**Comment by oli-obk** [comment]
Impl lgtm

r? @RalfJung

File: compiler/rustc_const_eval/src/transform/check_consts/resolver.rs [hunk 2]
```diff
@@ -265,6 +273,19 @@ pub(super) struct State {
...
+    fn clone(&self) -> Self {
+        State { qualif: self.qualif.clone(), borrow: self.borrow.clone() }
+    }
+
+    // Data flow engine when possible uses `clone_from` for domain values. Implementing it should
+    // avoid some intermediate memory allocations.
+    fn clone_from(&mut self, other: &Self) {
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/transform/check_consts/resolver.rs:283]: So this is not the default implementation generated by `derive`?

File: src/test/ui/mir-dataflow/indirect-mutation-offset.rs [hunk 3]
```diff
@@ -1,48 +0,0 @@
...
-}
-
-#[rustc_mir(rustc_peek_indirectly_mutable,stop_after_dataflow)]
-const BOO: i32 = {
-    let x = PartialInteriorMut {
-        zst: [],
-        cell: UnsafeCell::new(0),
-    };
-
-    let p_zst: *const _ = &x.zst ; // Doesn't cause `x` to get marked as indirectly mutable.
```
Comment (RalfJung) [inline src/test/ui/mir-dataflow/indirect-mutation-offset.rs:28]: The ZST is `Freeze`, so this test should behave as before, shouldn't it? Why is it being removed?

File: src/test/ui/mir-dataflow/indirect-mutation-offset.rs [hunk 4]
```diff
@@ -1,48 +0,0 @@
...
-
-    let rmut_cell = unsafe {
-        // Take advantage of the fact that `zst` and `cell` are at the same location in memory.
-        // This trick would work with any size type if miri implemented `ptr::offset`.
```
Comment (RalfJung) [inline src/test/ui/mir-dataflow/indirect-mutation-offset.rs:32]: FWIW, Miri (https://github.com/rust-lang/miri/) implements it, but the CTFE interpreter does not. Or rather, it didn't use to -- these days, `offset` is unstably `const fn`.

File: compiler/rustc_const_eval/src/transform/check_consts/resolver.rs (hunk 2 above)
Comment (RalfJung) [reply compiler/rustc_const_eval/src/transform/check_consts/resolver.rs:283]: Ah, that seems like an improvement worth making... but it has nothing to do with this PR, of course.

File: src/test/ui/mir-dataflow/indirect-mutation-offset.rs (hunk 3 above)
Comment (tmiasko) [reply src/test/ui/mir-dataflow/indirect-mutation-offset.rs:28]: The test requires `MaybeMutBorrowedLocals` data flow which no longer exists.

File: compiler/rustc_const_eval/src/transform/check_consts/resolver.rs (hunk 2 above)
Comment (tmiasko) [reply compiler/rustc_const_eval/src/transform/check_consts/resolver.rs]: Moved to a separate pull request.

File: src/test/ui/mir-dataflow/indirect-mutation-offset.rs (hunk 3 above)
Comment (RalfJung) [reply src/test/ui/mir-dataflow/indirect-mutation-offset.rs:28]: Ah okay. I don't know what this mir_peek stuff is doing anyway so I will just ignore this then. ;)

**Comment by RalfJung** [comment]
The analysis makes sense to me. The MaybeMutBorrowedLocals removal is code I have never seen but Oli reviewed it, so let's go ahead.

@bors r=RalfJung,oli-obk

**Pull request merged**